firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +590,325 @@ def parenthesize(arg, prec=None, parent=None):
+
+    ############
+    # TODO: preprocessing of gem for removing unneccesary component tensors
+    # print("not peprocessed",traversed_gem_expr_dag)
+    # traversed_gem_expr_dag = impero_utils.preprocess_traversedgem(traversed_gem_expr_dag[0])
+    # print("preprocessed",traversed_gem_expr_dag)
+    # print(traversed_gem_expr_dag[1])
+    # traversed_gem_expr_dag=traversed_gem_expr_dag[1]
+
+    # glue assignments to return variable
+    assignments = list(zip(ret_vars, traversed_gem_expr_dag))
+    impero_c = impero_utils.compile_gem(assignments, (), remove_zeros=False)
+
+    # Part B: impero_c to loopy
+    precision = 12
+    loopy_outer = generate_loopy(impero_c, args, precision, "double", "loopy_outer")
+    return loopy_outer
+
+
+# TODO: those should got into firedrake loopy!
+# STAGE 2c: register external function calls
+# the get_*_callable replaces the according callable
+# with the c-function which is defined in the preamble
+def get_inv_callable(loopy_merged):
+    class INVCallable(loopy.ScalarCallable):
+        def __init__(self, name, arg_id_to_dtype=None, 
+                     arg_id_to_descr=None, name_in_target=None):
+
+            super(INVCallable, self).__init__(name,
+                  arg_id_to_dtype=arg_id_to_dtype,
+                  arg_id_to_descr=arg_id_to_descr)
+
+            self.name = name
+            self.name_in_target = name_in_target
+
+        def with_types(self, arg_id_to_dtype, kernel, callables_table):
+            for i in range(0, len(arg_id_to_dtype)):
+                if i not in arg_id_to_dtype or arg_id_to_dtype[i] is None:
+                    # the types provided aren't mature enough to specialize the
+                    # callable
+                    return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                            callables_table)
+
+            mat_dtype = arg_id_to_dtype[0].numpy_dtype
+            name_in_target = "inverse_"
+
+            from loopy.types import NumpyType
+            return (self.copy(name_in_target=name_in_target,
+                              arg_id_to_dtype={0: NumpyType(mat_dtype), 1: NumpyType(int)}),
+                    callables_table)
+
+        def emit_call_insn(self, insn, target, expression_to_code_mapper):
+            assert self.is_ready_for_codegen()
+
+            assert isinstance(insn, loopy.CallInstruction)
+
+            parameters = insn.expression.parameters
+
+            parameters = list(parameters)
+            par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+            parameters.append(insn.assignees[0])
+            par_dtypes.append(self.arg_id_to_dtype[0])
+
+            from loopy.expression import dtype_to_type_context
+            from pymbolic.mapper.stringifier import PREC_NONE
+            from loopy.symbolic import SubArrayRef
+            from pymbolic import var
```

Comment:
Local imports have their uses, but they should only be used with deliberation.  Unless you have a good reason not to, place your imports at the beginning of the module (file).  If that made unclear where certain names come from (e.g. Is it a loo.py name? Is it GEM? Is it SLATE?), you can import the module instead (perhaps with an alias), or rename as you import.

Reply:
I was not quite sure if I leave the Callables in the compiler or if they should go into the firedrake branch of loopy.